lib/block-supports/layout.php

```diff
@@ -54,14 +55,25 @@ function gutenberg_get_layout_style( $selector, $layout, $has_block_gap_support
 		$wide_max_width_value = wp_strip_all_tags( explode( ';', $wide_max_width_value )[0] );
 
 		if ( $content_size || $wide_size ) {
-			$style  = "$selector > :where(:not(.alignleft):not(.alignright)) {";
+			$style  = "$selector > :where(:not(.alignleft):not(.alignright):not(.alignfull)) {";
 			$style .= 'max-width: ' . esc_html( $all_max_width_value ) . ';';
 			$style .= 'margin-left: auto !important;';
 			$style .= 'margin-right: auto !important;';
 			$style .= '}';
 
 			$style .= "$selector > .alignwide { max-width: " . esc_html( $wide_max_width_value ) . ';}';
 			$style .= "$selector .alignfull { max-width: none; }";
+
+			// Handle negative margins for alignfull children of blocks with custom padding set.
+			// They're added separately because padding might only be set on one side.
+			if ( isset( $block_padding['right'] ) ) {
+				$padding_right = $block_padding['right'];
+				$style        .= "$selector > .alignfull { margin-right: -$padding_right; }";
```

Comment:
Oh, I just thought of something — as of https://github.com/WordPress/gutenberg/pull/41527 and https://github.com/WordPress/gutenberg/pull/41990 this padding value might contain a not-immediately-usable value like `var:preset|spacing|70` and so will possibly need to be parsed in order to be usable / constructed here.

In the style engine, the CSS var is parsed into a `var()` value, so I wonder if we need to call the style engine or similar method here so that this plays nicely with spacing presets that might be stored in a `style.spacing.padding` block attribute?

Sorry I didn't think of this earlier!

Here's the style engine method that converts the value to a `var()` value: https://github.com/WordPress/gutenberg/blob/dd8d577b859a07d1ea1798104123a162311435de/packages/style-engine/class-wp-style-engine.php#L263

I wonder if we can do something sorta similar to the spacing block support and construct an array with the `'margin'` values we want, then call the style engine like [on this line](https://github.com/WordPress/gutenberg/blob/dd8d577b859a07d1ea1798104123a162311435de/lib/block-supports/spacing.php#L60), and then inject the returned string between the braces, e.g. `"$selector > .alignfull {  $generated_string_from_the_style_engine }"`?

Reply:
Ooooh thanks for the heads up, I'll look into it! This complicates things a little 😅

Reply:
Hopefully only slightly! 🤞

Happy to give you a hand with it on Monday if it looks at all confusing, as @ramonjd and I have looked at a few of these sorts of fiddly cases 🙂

Reply:
I'm not quite sure how to make those padding values contain presets - from the description on #41527 it sounds like they should already be available in core, but I can't work out how to make the custom padding controls on the Group block use those presets if that makes sense?

Reply:
> I can't work out how to make the custom padding controls on the Group block use those presets if that makes sense?

It does make sense — the UI for inserting those CSS variables hasn't been shipped yet (@glendaviesnz is in the process of building it) so there aren't any blocks that have that value _yet_, but the expectation is that there will be quite soon.

In the interim, we should be able to use the code editor view to manually enter values. The following markup works for me when copy / pasted into the code view:

```
<!-- wp:group {"style":{"spacing":{"padding":{"top":"var:preset|spacing|70","right":"var:preset|spacing|50","bottom":"var:preset|spacing|70","left":"var:preset|spacing|50"}}},"backgroundColor":"vivid-red"} -->
<div class="wp-block-group has-vivid-red-background-color has-background" style="padding-top:var(--wp--preset--spacing--70);padding-right:var(--wp--preset--spacing--50);padding-bottom:var(--wp--preset--spacing--70);padding-left:var(--wp--preset--spacing--50)"><!-- wp:paragraph -->
<p>A paragraph</p>
<!-- /wp:paragraph --></div>
<!-- /wp:group -->
```

Reply:
Ahhhh, thanks, I can reproduce it now!

Reply:
I've pushed a fix for this, but I had to change `get_block_supports_styles` in the style engine class to accept an option to output the CSS as an array instead of a declaration string. Hopefully that's ok! [Here's the changeset](https://github.com/WordPress/gutenberg/pull/42085/commits/74fef96014b56c9a3c3e0df4270eac7e8ab3c944).

Reply:
Oh, interesting problem, apologies I missed that the API there (and how the value is used in this support) makes it a bit harder to work with. I wonder if an alternative to changing the return value of `css` (which for predictability might be better to keep as a string) if it'd be helpful to return a `'declarations'` key in the array returned by `get_block_supports_styles`. I'll just CC: @ramonjd too, since he's been working a lot on the API for block supports and the style engine lately 🙂

If it's too fiddly or if it feels like it's holding up this PR, we could always look at fixing this up in a follow-up PR since the UI for using the presets hasn't landed yet.

Reply:
Shouldn't be hard to adjust that. Would we still need the option in that case, or should we just return `declarations` by default?

Reply:
I'd lean toward returning by default, personally, to keep it simpler?

Reply:
Done!

Reply:
> I wonder if an alternative to changing the return value of css (which for predictability might be better to keep as a string) 

Yeah, I think it'd be neater to keep the return value as is, despite the options flag. I can't think of a concrete case, but I suspect mixing values will complicate things down the track. 

>  I wonder if an alternative to changing the return value of css (which for predictability might be better to keep as a string) if it'd be helpful to return a 'declarations' key in the array returned by get_block_supports_styles

That's a valid option, and the reason why I started off with an array here.

In fact, it's so compelling, I had to try it out: 

https://github.com/WordPress/gutenberg/pull/42307

I had explored dropping the `[ 'css' => ..., 'classnames' => ... ]` return value [in this test PR](https://github.com/WordPress/gutenberg/pull/42144), but decided that it might be very useful for when we look at generating semantic classnames from block-specific styles at the block support level, e.g., `has-padding-10px` or whatever.

If https://github.com/WordPress/gutenberg/pull/42307 is okay, we can merge and then... merge 😄

Reply:
> If https://github.com/WordPress/gutenberg/pull/42307 is okay, we can merge and then... merge 😄

We'll likely need to merge / rebase the upstream Layout PR (https://github.com/WordPress/gutenberg/pull/40875), too. I think that one's ready for a final review now, though, so 🤞 it shouldn't be too much longer before this PR can then be pointed to `trunk`?

Reply:
> We'll likely need to merge / rebase the upstream Layout PR (#40875), too

I was just thinking that! I'll have another look at it, I think it's pretty much ready for ✅ 

Also thanks for #42307 @ramonjd!